iceoryx_posh/source/popo/active_call_set.cpp

```diff
@@ -0,0 +1,223 @@
+    : ActiveCallSet(runtime::PoshRuntime::getInstance().getMiddlewareEventVariable())
+{
+}
+
+ActiveCallSet::ActiveCallSet(EventVariableData* eventVariable) noexcept
+    : m_eventVariable(eventVariable)
+    , m_eventListener(*eventVariable)
+{
+    m_thread = std::thread(&ActiveCallSet::threadLoop, this);
+}
+
+ActiveCallSet::~ActiveCallSet()
+{
+    m_wasDtorCalled.store(true, std::memory_order_relaxed);
+    m_eventListener.destroy();
+
+    m_thread.join();
+    m_eventVariable->m_toBeDestroyed.store(true, std::memory_order_relaxed);
+}
+
+cxx::expected<uint32_t, ActiveCallSetError>
+ActiveCallSet::addEvent(void* const origin,
+                        const uint64_t eventType,
+                        const uint64_t eventTypeHash,
+                        CallbackRef_t<void> callback,
+                        TranslationCallbackRef_t translationCallback,
+                        const cxx::MethodCallback<void, uint64_t> invalidationCallback) noexcept
+{
+    std::lock_guard<std::mutex> lock(m_addEventMutex);
+
+    for (uint32_t i = 0U; i < MAX_NUMBER_OF_EVENTS_PER_ACTIVE_CALL_SET; ++i)
+    {
+        if (m_events[i]->isEqualTo(origin, eventType, eventTypeHash))
+        {
+            return cxx::error<ActiveCallSetError>(ActiveCallSetError::EVENT_ALREADY_ATTACHED);
+        }
+    }
+
+    uint32_t index = 0U;
+    if (!m_indexManager.pop(index))
+    {
+        return cxx::error<ActiveCallSetError>(ActiveCallSetError::ACTIVE_CALL_SET_FULL);
+    }
+
+    m_events[index]->init(index, origin, eventType, eventTypeHash, callback, translationCallback, invalidationCallback);
+    return cxx::success<uint32_t>(index);
+}
+
+void ActiveCallSet::removeEvent(void* const origin, const uint64_t eventType, const uint64_t eventTypeHash) noexcept
+{
+    for (uint32_t index = 0U; index < MAX_NUMBER_OF_EVENTS_PER_ACTIVE_CALL_SET; ++index)
+    {
+        if (m_events[index]->resetIfEqualTo(origin, eventType, eventTypeHash))
+        {
+            m_indexManager.push(index);
+            break;
+        }
+    }
+}
+
+uint64_t ActiveCallSet::size() const noexcept
+{
+    return m_indexManager.indicesInUse();
+}
+
+void ActiveCallSet::threadLoop() noexcept
+{
+    while (m_wasDtorCalled.load(std::memory_order_relaxed) == false)
+    {
+        auto activateNotificationIds = m_eventListener.wait();
+
+        cxx::forEach(activateNotificationIds, [this](auto id) { m_events[id]->executeCallback(); });
+    }
+}
+
+void ActiveCallSet::removeTrigger(const uint64_t index) noexcept
+{
+    if (index >= MAX_NUMBER_OF_EVENTS_PER_ACTIVE_CALL_SET)
+    {
+        return;
+    }
+
+    if (m_events[index]->reset())
+    {
+        m_indexManager.push(static_cast<uint32_t>(index));
+    }
+}
+
+////////////////
+// Event_t
+////////////////
+ActiveCallSet::Event_t::~Event_t()
+{
+    reset();
+}
+
+void ActiveCallSet::Event_t::executeCallback() noexcept
+{
+    if (!isInitialized())
+    {
+        return;
+    }
+
+    m_translationCallback(m_origin, m_callback);
```

Comment:
what happens if the Event is invalidated after `isInitialized` is called but before executing the callback?

Reply:
This can never happen since the `Event_t` class is not threadsafe and is only handled in single threaded context. The `smart_lock` ensures that this does not happen in a multithreaded context.